File: fore-core/src/main/java/co/early/fore/core/observer/ObservableGroupImp.java
package co.early.fore.core.observer;

import java.util.Arrays;
import java.util.List;

import co.early.fore.core.Affirm;

public class ObservableGroupImp implements ObservableGroup {

    private final List<Observable> observablesList;

    public ObservableGroupImp(Observable... observablesList) {
        this.observablesList = Arrays.asList(Affirm.notNull(observablesList));
        checkObservables();
    }

    @Override
    public void addObserver(Observer observer) {
        for (Observable observable : observablesList) {
            observable.addObserver(observer);
        }
    }

    @Override
    public void removeObserver(Observer observer) {
        for (Observable observable : observablesList) {
            observable.removeObserver(observer);
        }
    }

    private void checkObservables() {
        for (Observable observable : observablesList) {
            if (observable == null) {
                throw new RuntimeException("ObservableGroup has been instantiated with at least one null observable");
            }
        }
    }
}

File: fore-core-kt/src/main/java/co/early/fore/kt/core/observer/ObservableGroupImp.kt
package co.early.fore.kt.core.observer

import co.early.fore.core.observer.ObservableGroup
import co.early.fore.core.observer.Observable
import co.early.fore.core.observer.Observer

class ObservableGroupImp(vararg observablesList: Observable) : ObservableGroup {

    private val observablesList: List<Observable> = listOf(*observablesList)

    override fun addObserver(observer: Observer) {
        for (observable in observablesList) {
            observable.addObserver(observer)
        }
    }

    override fun removeObserver(observer: Observer) {
        for (observable in observablesList) {
            observable.removeObserver(observer)
        }
    }
}